In [1]:
from pyspark.sql import SparkSession

spark = SparkSession.builder \
    .appName("SpotifyEDA") \
    .getOrCreate()

Starting Spark application


ID,YARN Application ID,Kind,State,Spark UI,Driver log,Current session?
0,application_1784461108585_0001,pyspark,idle,Link,Link,✔


SparkSession available as 'spark'.


In [2]:
from pyspark.sql.types import *

In [6]:
songs_schema = StructType([
    StructField("track_uri", StringType(), False),
    StructField("track_name", StringType(), True),
    StructField("artist_names", StringType(), True),
    StructField("artist_uris", StringType(), True),
    StructField("label", StringType(), True),
    StructField("release_date", DateType(), True),
    StructField("all_uris", StringType(), True)
])

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [7]:
artists_schema = StructType([
    StructField("artist_uri", StringType(), False),
    StructField("artist_name", StringType(), True)
])

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [8]:
albums_schema = StructType([
    StructField("album_uri", StringType(), False),
    StructField("album_name", StringType(), True),
    StructField("artist_names", StringType(), True),
    StructField("artist_uris", StringType(), True),
    StructField("label", StringType(), True),
    StructField("release_date", DateType(), True)
])

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [9]:
artwork_schema = StructType([
    StructField("uri", StringType(), False),
    StructField("type", StringType(), True),
    StructField("image_url", StringType(), True),
    StructField("last_seen_date", DateType(), True)
])

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [10]:
links_schema = StructType([
    StructField("spotify_uri", StringType(), False),
    StructField("type", StringType(), True),
    StructField("youtube_music_id", StringType(), True),
    StructField("youtube_video_id", StringType(), True),
    StructField("youtube_browse_id", StringType(), True),
    StructField("youtube_channel_id", StringType(), True)
])

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [3]:
charts_songs_schema = StructType([
    StructField("date", DateType(), True),
    StructField("country", StringType(), True),
    StructField("rank", IntegerType(), True),
    StructField("uri", StringType(), False),
    StructField("artist_names", StringType(), True),
    StructField("track_name", StringType(), True),
    StructField("label", StringType(), True),
    StructField("peak_rank", IntegerType(), True),
    StructField("previous_rank", IntegerType(), True),
    StructField("days_on_chart", IntegerType(), True),
    StructField("streams", LongType(), True),
    StructField("consecutive_days", IntegerType(), True),
    StructField("entry_status", StringType(), True),
    StructField("peak_date", DateType(), True),
    StructField("entry_rank", IntegerType(), True),
    StructField("entry_date", DateType(), True),
    StructField("release_date", DateType(), True),
    StructField("artist_uris", StringType(), True)
])

In [12]:
charts_artists_schema = StructType([
    StructField("date", DateType(), True),
    StructField("country", StringType(), True),
    StructField("rank", IntegerType(), True),
    StructField("uri", StringType(), False),
    StructField("artist_name", StringType(), True),
    StructField("peak_rank", IntegerType(), True),
    StructField("previous_rank", IntegerType(), True),
    StructField("days_on_chart", IntegerType(), True),
    StructField("consecutive_days", IntegerType(), True),
    StructField("entry_status", StringType(), True),
    StructField("peak_date", DateType(), True),
    StructField("entry_rank", IntegerType(), True),
    StructField("entry_date", DateType(), True)
])

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [13]:
charts_albums_schema = StructType([
    StructField("date", DateType(), True),
    StructField("country", StringType(), True),
    StructField("rank", IntegerType(), True),
    StructField("uri", StringType(), False),
    StructField("album_name", StringType(), True),
    StructField("artist_names", StringType(), True),
    StructField("label", StringType(), True),
    StructField("peak_rank", IntegerType(), True),
    StructField("previous_rank", IntegerType(), True),
    StructField("weeks_on_chart", IntegerType(), True),
    StructField("consecutive_weeks", IntegerType(), True),
    StructField("entry_status", StringType(), True),
    StructField("peak_date", DateType(), True),
    StructField("entry_rank", IntegerType(), True),
    StructField("entry_date", DateType(), True),
    StructField("release_date", DateType(), True),
    StructField("artist_uris", StringType(), True)
])

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [14]:
songs_df = spark.read \
    .schema(songs_schema) \
    .option("header", "true") \
    .option("dateFormat", "yyyy-MM-dd") \
    .csv("s3://group-1-dbda/raw_data/songs.csv")

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [15]:
artists_df = spark.read \
    .schema(artists_schema) \
    .option("header", "true") \
    .csv("s3://group-1-dbda/raw_data/artists.csv")

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [16]:
albums_df = spark.read \
    .schema(albums_schema) \
    .option("header", "true") \
    .option("dateFormat", "yyyy-MM-dd") \
    .csv("s3://group-1-dbda/raw_data/albums.csv")

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [17]:
artwork_df = spark.read \
    .schema(artwork_schema) \
    .option("header", "true") \
    .option("dateFormat", "yyyy-MM-dd") \
    .csv("s3://group-1-dbda/raw_data/artwork.csv")

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [18]:
links_df = spark.read \
    .schema(links_schema) \
    .option("header", "true") \
    .csv("s3://group-1-dbda/raw_data/links.csv")

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [4]:
charts_songs_df = spark.read.parquet(
    "s3://group-1-dbda/raw_data/charts_songs_daily.parquet"
)

In [8]:
my_bucket = "rafat-spotify-medallion"
bronze_path = "s3://" + my_bucket + "/bronze/charts_songs_daily/"

In [9]:
source_path =  "s3://group-1-dbda/raw_data/charts_songs_daily.parquet/"

print(source_path)

s3://group-1-dbda/raw_data/charts_songs_daily.parquet/

In [10]:
charts_songs_df.write \
    .mode("overwrite") \
    .parquet(bronze_path)

In [11]:
bronze_df = spark.read.parquet(
    "s3://rafat-spotify-medallion/bronze/charts_songs_daily/"
)

In [12]:
single_df = bronze_df.coalesce(1)

In [13]:
single_df.write \
    .mode("overwrite") \
    .parquet("s3://rafat-spotify-medallion/download/charts_songs_daily_single/")

In [22]:
charts_artists_df = spark.read.parquet(
    "s3://group-1-dbda/raw_data/charts_artists_daily.parquet"
)

In [ ]:
charts_songs_df.write \
    .mode("overwrite") \
    .parquet(bronze_path)

In [23]:
charts_albums_df= spark.read \
    .schema(charts_albums_schema) \
    .option("header", "true") \
    .option("dateFormat", "yyyy-MM-dd") \
    .csv("s3://group-1-dbda/raw_data/charts_albums_weekly.csv")

In [29]:
songs_df.show(20, truncate=False)

+------------------------------------+-----------------------------------------------+---------------------------------------------+-----------------------------------------------------------------------------------------------------------------+---------------------------------------+------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|track_uri                           |track_name                                     |artist_names                                 |artist_uris                                                                                                      |label                                  |release_date|all_uris                                                                                                                                                                                |
+-----------------------------

In [30]:
albums_df.show(20, truncate=False)

+------------------------------------+-----------------------------+------------------------+---------------------------------------------------------------------------+---------------------------------------+------------+
|album_uri                           |album_name                   |artist_names            |artist_uris                                                                |label                                  |release_date|
+------------------------------------+-----------------------------+------------------------+---------------------------------------------------------------------------+---------------------------------------+------------+
|spotify:album:3fq06DDIkadDRB9RuBb22v|Afrodisíaco                  |Rauw Alejandro          |spotify:artist:1mcTU81TzQhprhouKaTkpq                                      |Sony Music Latin/Duars Entertainment   |2020-11-13  |
|spotify:album:0p2yf6DucEgvj8Uk8KXJJv|PAPI JUANCHO                 |Maluma                  |spotify:artist:

In [31]:
artists_df.show(20, truncate=False)

+-------------------------------------+------------------+
|artist_uri                           |artist_name       |
+-------------------------------------+------------------+
|spotify:artist:3TVXtAsR1Inumwj472S9r4|Drake             |
|spotify:artist:1UwTwoC4T1i6vzwsQgIWB0|Feza              |
|spotify:artist:1RyvyyTE3xzB2ZywiAwp0i|Future            |
|spotify:artist:1zEHBw7xQf0drXZagRkciU|Molly Santana     |
|spotify:artist:1URnnhqYAYcrqrcwql10ft|21 Savage         |
|spotify:artist:2zMD4U9OQAR3xuq6cjer8p|Calvin Fallo      |
|spotify:artist:0ysbLY7TUvj3AKV2W7ZCFY|LIMIT NALA        |
|spotify:artist:2HPaUgqeutzr3jx5a9WyDV|PARTYNEXTDOOR     |
|spotify:artist:5iEByMj7fW1lewiS4Ik91v|Trechyson Molly vx|
|spotify:artist:0fqv3yhjGZJ20OUjtIxj9H|DJ Promatic SA    |
|spotify:artist:3lCc4C1DNASVQR53kwxGue|Dendofela         |
|spotify:artist:0HX88eTlictNvgUjBLn03r|Ba Bethe Gashoazen|
|spotify:artist:4I8qIZwLyx6CqE88hHGnjJ|Nova sa style     |
|spotify:artist:4feLzuIrjiKM2KirRK56Zz|Westboy          

In [32]:
artwork_df.show(5, truncate=False)

+------------------------------------+----+----------------------------------------------------------------+--------------+
|uri                                 |type|image_url                                                       |last_seen_date|
+------------------------------------+----+----------------------------------------------------------------+--------------+
|spotify:track:000N4CJL8IjQ0f2I4grgBO|song|https://i.scdn.co/image/ab67616d00001e02e96f1ee613079b5845fe84bb|2023-05-18    |
|spotify:track:000RW47rhEkSqjgTrZx7YX|song|https://i.scdn.co/image/ab67616d00001e025c62203c9a356ed11c94f4c7|2021-10-21    |
|spotify:track:000iCrF4KRplQsmtCs7mR1|song|https://i.scdn.co/image/ab67616d00001e023234a138288b9d81cc8377a1|2026-02-15    |
|spotify:track:000n6Lx4yqUAslF1x3JeFY|song|https://i.scdn.co/image/ab67616d00001e022643ebfdf951fddf1bc3dca7|2024-09-05    |
|spotify:track:000xQL6tZNLJzIrtIgxqSl|song|https://i.scdn.co/image/ab67616d00001e02eadac26cea27137ae556e21e|2017-11-29    |
+-------

In [33]:
links_df.show(5, truncate=False)

+-------------------------------------+------+----------------+----------------+------------------------+------------------+
|spotify_uri                          |type  |youtube_music_id|youtube_video_id|youtube_browse_id       |youtube_channel_id|
+-------------------------------------+------+----------------+----------------+------------------------+------------------+
|spotify:artist:00FQb4jTyendYWaN8pK0wa|artist|null            |null            |UCyqq-aiu3vEHuf5NhwmOJcw|null              |
|spotify:artist:00x1fYSGhdqScXBRpSj3DW|artist|null            |null            |UC6GBKGYX6b3guHlqHLx6IzQ|null              |
|spotify:artist:02gSuSAWEdWa5UOvqzjX6v|artist|null            |null            |UCf5GUgSQC4CU8idz_OH8jqQ|null              |
|spotify:artist:04gDigrS5kc9YWfZHwBETP|artist|null            |null            |UCdFe4KkWwZ_twpo-UECR-Nw|null              |
|spotify:artist:06HL4z0CvFAxyc27GXpf02|artist|null            |null            |UCPC0L1d253x-KuMNwa05TpA|null              |


In [24]:
charts_songs_df.show(5, truncate=False)

+----------+-------+----+------------------------------------+----------------------------------------+---------------------------------------------+---------------------------+---------+-------------+-------------+-------+----------------+------------+----------+----------+----------+------------+---------------------------------------------------------------------------+
|date      |country|rank|uri                                 |artist_names                            |track_name                                   |label                      |peak_rank|previous_rank|days_on_chart|streams|consecutive_days|entry_status|peak_date |entry_rank|entry_date|release_date|artist_uris                                                                |
+----------+-------+----+------------------------------------+----------------------------------------+---------------------------------------------+---------------------------+---------+-------------+-------------+-------+----------------+--------

In [25]:
charts_artists_df.show(5, truncate=False)

+----------+-------+----+-------------------------------------+-------------+---------+-------------+-------------+----------------+------------+----------+----------+----------+
|date      |country|rank|uri                                  |artist_name  |peak_rank|previous_rank|days_on_chart|consecutive_days|entry_status|peak_date |entry_rank|entry_date|
+----------+-------+----+-------------------------------------+-------------+---------+-------------+-------------+----------------+------------+----------+----------+----------+
|2021-10-21|ae     |1   |spotify:artist:3TVXtAsR1Inumwj472S9r4|Drake        |1        |-1           |1            |1               |NEW_ENTRY   |2021-10-21|1         |2021-10-21|
|2021-10-21|ae     |2   |spotify:artist:1uNFoZAHBGtllmzznpCI3s|Justin Bieber|2        |-1           |1            |1               |NEW_ENTRY   |2021-10-21|2         |2021-10-21|
|2021-10-21|ae     |3   |spotify:artist:3Nrfpe0tUJi4K4DXYWgMUX|BTS          |3        |-1           |1   

In [45]:
charts_albums_df.show(5, truncate=False)

+----------+-------+----+------------------------------------+---------------+--------------+------------------------------------+---------+-------------+--------------+-----------------+------------+----------+----------+----------+------------+-------------------------------------+
|date      |country|rank|uri                                 |album_name     |artist_names  |label                               |peak_rank|previous_rank|weeks_on_chart|consecutive_weeks|entry_status|peak_date |entry_rank|entry_date|release_date|artist_uris                          |
+----------+-------+----+------------------------------------+---------------+--------------+------------------------------------+---------+-------------+--------------+-----------------+------------+----------+----------+----------+------------+-------------------------------------+
|2020-11-19|ar     |1   |spotify:album:3fq06DDIkadDRB9RuBb22v|Afrodisíaco    |Rauw Alejandro|Sony Music Latin/Duars Entertainment|1        |-1   

In [46]:
datasets = {
    "Songs": songs_df,
    "Artists": artists_df,
    "Albums": albums_df,
    "Artwork": artwork_df,
    "Links": links_df,
    "Charts Songs": charts_songs_df,
    "Charts Artists": charts_artists_df,
    "Charts Albums": charts_albums_df
}

for name, df in datasets.items():
    print("=" * 60)
    print(name)
    print("Rows:", df.count())
    print("Columns:", len(df.columns))

Charts Artists
('Rows:', 22962712)
('Columns:', 13)
Charts Albums
('Rows:', 2795375)
('Columns:', 17)
Charts Songs
('Rows:', 42757018)
('Columns:', 18)
Links
('Rows:', 101420)
('Columns:', 6)
Artists
('Rows:', 63775)
('Columns:', 2)
Albums
('Rows:', 24675)
('Columns:', 6)
Artwork
('Rows:', 295712)
('Columns:', 4)
Songs
('Rows:', 207164)
('Columns:', 7)

In [47]:
for name, df in datasets.items():
    print(name, df.rdd.getNumPartitions())

('Charts Artists', 2)
('Charts Albums', 5)
('Charts Songs', 6)
('Links', 2)
('Artists', 1)
('Albums', 1)
('Artwork', 2)
('Songs', 2)

In [48]:
songs_df.describe().show()

+-------+--------------------+----------+--------------------+-----------------+--------------------+--------------------+
|summary|           track_uri|track_name|        artist_names|      artist_uris|               label|            all_uris|
+-------+--------------------+----------+--------------------+-----------------+--------------------+--------------------+
|  count|              207164|    207163|              207163|           207164|              207161|              207164|
|   mean|                null|  Infinity|   962.7340425531914|             null| 3.364782189056604E8|                null|
| stddev|                null|       NaN|  1669.0572522540267|             null| 8.124353959123614E9|                null|
|    min|spotify:track:000...|         !| ""Побєда"" і Бер...| Q-Tip & Denaun)"|"""Hitmen Music /...|          1983-01-01|
|    max|spotify:track:7zz...|        🪐|              하숙자|     БЕЗ ОБМЕЖЕНЬ|            피네이션|spotify:track:7zz...|
+-------+---------------

In [50]:
from pyspark.sql.functions import *

songs_df.select([
    count(when(col(c).isNull(), c)).alias(c)
    for c in songs_df.columns
]).show()

+---------+----------+------------+-----------+-----+------------+--------+
|track_uri|track_name|artist_names|artist_uris|label|release_date|all_uris|
+---------+----------+------------+-----------+-----+------------+--------+
|       23|        24|          24|         23|   26|        2631|      23|
+---------+----------+------------+-----------+-----+------------+--------+

In [51]:
#percentage
total = songs_df.count()

for c in songs_df.columns:
    songs_df.select(
        (
            count(when(col(c).isNull(), c))/total*100
        ).alias(c)
    ).show()

+---------+
|track_uri|
+---------+
|      0.0|
+---------+

+--------------------+
|          track_name|
+--------------------+
|4.827093510455485E-4|
+--------------------+

+--------------------+
|        artist_names|
+--------------------+
|4.827093510455485E-4|
+--------------------+

+-----------+
|artist_uris|
+-----------+
|        0.0|
+-----------+

+--------------------+
|               label|
+--------------------+
|0.001448128053136...|
+--------------------+

+-----------------+
|     release_date|
+-----------------+
|1.270008302600838|
+-----------------+

+--------+
|all_uris|
+--------+
|     0.0|
+--------+

In [53]:
from pyspark.sql.functions import *

artists_df.select([
    count(when(col(c).isNull(), c)).alias(c)
    for c in artists_df.columns
]).show()

+----------+-----------+
|artist_uri|artist_name|
+----------+-----------+
|         0|          0|
+----------+-----------+

In [54]:
from pyspark.sql.functions import *

albums_df.select([
    count(when(col(c).isNull(), c)).alias(c)
    for c in albums_df.columns
]).show()

+---------+----------+------------+-----------+-----+------------+
|album_uri|album_name|artist_names|artist_uris|label|release_date|
+---------+----------+------------+-----------+-----+------------+
|       13|        14|          14|         13|   15|         768|
+---------+----------+------------+-----------+-----+------------+

In [26]:
from pyspark.sql.functions import *

charts_songs_df.select([
    count(when(col(c).isNull(), c)).alias(c)
    for c in charts_songs_df.columns
]).show()

+----+-------+----+---+------------+----------+-----+---------+-------------+-------------+-------+----------------+------------+---------+----------+----------+------------+-----------+
|date|country|rank|uri|artist_names|track_name|label|peak_rank|previous_rank|days_on_chart|streams|consecutive_days|entry_status|peak_date|entry_rank|entry_date|release_date|artist_uris|
+----+-------+----+---+------------+----------+-----+---------+-------------+-------------+-------+----------------+------------+---------+----------+----------+------------+-----------+
|   0|      0|   0|  0|       44711|     44711|42993|        0|            0|            0|      0|               0|           0|        0|         0|         0|      636254|          0|
+----+-------+----+---+------------+----------+-----+---------+-------------+-------------+-------+----------------+------------+---------+----------+----------+------------+-----------+

In [27]:
# Duplicate Analysis(charts_songs_daily)
# Duplicate Analysis - Charts Songs

total_rows = charts_songs_df.count()
distinct_rows = charts_songs_df.distinct().count()
duplicate_rows = total_rows - distinct_rows

if total_rows > 0:
    duplicate_percent = (duplicate_rows * 100.0) / total_rows
else:
    duplicate_percent = 0

print("=" * 60)
print("Duplicate Analysis - Charts Songs")
print("=" * 60)
print("Total Rows      :", total_rows)
print("Distinct Rows   :", distinct_rows)
print("Duplicate Rows  :", duplicate_rows)
print("Duplicate %     :", duplicate_percent)

Duplicate Analysis - Charts Songs
('Total Rows      :', 42757018)
('Distinct Rows   :', 42757018)
('Duplicate Rows  :', 0)
('Duplicate %     :', 0.0)

**No exact duplicate records were identified in the charts_songs_df dataset. Each row represents a unique chart observation (based on the combination of date, country, song, and ranking information). This indicates that the dataset maintains record-level integrity and does not require duplicate removal before downstream ETL processing.**

In [33]:
from pyspark.sql.functions import *

charts_songs_df.select([
    count(when(col(c).isNull(), c)).alias(c)
    for c in charts_songs_df.columns
]).show()

+----+-------+----+---+------------+----------+-----+---------+-------------+-------------+-------+----------------+------------+---------+----------+----------+------------+-----------+
|date|country|rank|uri|artist_names|track_name|label|peak_rank|previous_rank|days_on_chart|streams|consecutive_days|entry_status|peak_date|entry_rank|entry_date|release_date|artist_uris|
+----+-------+----+---+------------+----------+-----+---------+-------------+-------------+-------+----------------+------------+---------+----------+----------+------------+-----------+
|   0|      0|   0|  0|       44711|     44711|42993|        0|            0|            0|      0|               0|           0|        0|         0|         0|      636254|          0|
+----+-------+----+---+------------+----------+-----+---------+-------------+-------------+-------+----------------+------------+---------+----------+----------+------------+-----------+

| Column         | Missing Values | Interpretation                                                                                                                   |
| -------------- | -------------: | -------------------------------------------------------------------------------------------------------------------------------- |
| `release_date` |        636,254 | Some songs do not have release date metadata available. Useful for lifecycle analysis, but not critical for chart/rank analysis. |
| `artist_names` |         44,711 | Artist name missing for a small portion of records.                                                                              |
| `track_name`   |         44,711 | Track name missing for same rows as artist names, likely metadata lookup issue.                                                  |
| `label`        |         42,993 | Label metadata unavailable for some records.                                                                                     |
| Core fields    |              0 | `date`, `country`, `rank`, `uri`, `streams`, chart dates, and ranking fields are complete.                                       |


**The core analytical fields are complete, so stream/rank/country/date-based analysis is reliable. Missing values are mostly in descriptive metadata columns, which should be handled during ETL by keeping the records and filling metadata fields with values like "Unknown" where required.**

In [34]:
#missing percentage
total_rows = charts_songs_df.count()

charts_songs_missing_percent = charts_songs_df.select([
    (
        (count(when(col(c).isNull(), c)) / total_rows) * 100
    ).alias(c)
    for c in charts_songs_df.columns
])

charts_songs_missing_percent.show()

+----+-------+----+---+------------------+------------------+-------------------+---------+-------------+-------------+-------+----------------+------------+---------+----------+----------+------------------+-----------+
|date|country|rank|uri|      artist_names|        track_name|              label|peak_rank|previous_rank|days_on_chart|streams|consecutive_days|entry_status|peak_date|entry_rank|entry_date|      release_date|artist_uris|
+----+-------+----+---+------------------+------------------+-------------------+---------+-------------+-------------+-------+----------------+------------+---------+----------+----------+------------------+-----------+
| 0.0|    0.0| 0.0|0.0|0.1045699679056196|0.1045699679056196|0.10055191407408254|      0.0|          0.0|          0.0|    0.0|             0.0|         0.0|      0.0|       0.0|       0.0|1.4880691632891705|        0.0|
+----+-------+----+---+------------------+------------------+-------------------+---------+-------------+-----------

| Column               |  Missing % | Interpretation                                                                                                          |
| -------------------- | ---------: | ----------------------------------------------------------------------------------------------------------------------- |
| `release_date`       |  **1.49%** | Very few songs are missing release date metadata. This is acceptable and will have minimal impact on temporal analyses. |
| `artist_names`       | **0.105%** | Negligible missing values.                                                                                              |
| `track_name`         | **0.105%** | Negligible missing values.                                                                                              |
| `label`              | **0.101%** | Very small amount of missing label information.                                                                         |
| Remaining 14 columns |     **0%** | Complete with no missing values.                                                                                        |


The dataset demonstrates high data completeness.

The core analytical attributes including:

Date
Country
Rank
URI
Streams
Peak Rank
Previous Rank
Days on Chart
Consecutive Days

contain 100% complete data, making the dataset highly reliable for trend analysis, ranking analysis, market intelligence, and dashboard reporting.

The missing values are confined to metadata fields (artist_names, track_name, label, and release_date) and represent less than 1.5% of the dataset. These missing values are unlikely to affect business analytics significantly.

In [35]:
#step 3: Unique value Analysis
from pyspark.sql.functions import approx_count_distinct

charts_songs_df.select([
    approx_count_distinct(c).alias(c)
    for c in charts_songs_df.columns
]).show(truncate=False)

+----+-------+----+------+------------+----------+-----+---------+-------------+-------------+-------+----------------+------------+---------+----------+----------+------------+-----------+
|date|country|rank|uri   |artist_names|track_name|label|peak_rank|previous_rank|days_on_chart|streams|consecutive_days|entry_status|peak_date|entry_rank|entry_date|release_date|artist_uris|
+----+-------+----+------+------------+----------+-----+---------+-------------+-------------+-------+----------------+------------+---------+----------+----------+------------+-----------+
|3205|71     |195 |256793|91691       |177031    |27625|195      |195          |3132         |1177390|3132            |5           |3205     |195       |3205      |8430        |97427      |
+----+-------+----+------+------------+----------+-----+---------+-------------+-------------+-------+----------------+------------+---------+----------+----------+------------+-----------+

| Column           | Approx. Distinct Values | Interpretation                                                                                      |
| ---------------- | ----------------------: | --------------------------------------------------------------------------------------------------- |
| date             |               **3,205** | Approximately 8.8 years of daily chart data (2017–2026).                                            |
| country          |                  **71** | Spotify chart data covers 71 countries/regions, providing strong global market coverage.            |
| rank             |                 **195** | Each daily chart contains ranks from 1–200 (approximation gives 195).                               |
| uri              |             **256,793** | Unique Spotify track identifiers. High-cardinality identifier.                                      |
| artist_names     |              **91,691** | Large number of unique artists represented in the dataset.                                          |
| track_name       |             **177,031** | Extensive catalog of unique songs.                                                                  |
| label            |              **27,625** | High diversity of record labels and distributors.                                                   |
| peak_rank        |                 **195** | Peak ranking ranges from 1–200.                                                                     |
| previous_rank    |                 **195** | Previous ranking values also span the chart range.                                                  |
| days_on_chart    |               **3,132** | Songs remain on charts for varying durations, indicating long-term popularity analysis is possible. |
| streams          |           **1,177,390** | Extremely high cardinality; continuous numerical metric.                                            |
| consecutive_days |               **3,132** | Supports retention and longevity analysis.                                                          |
| entry_status     |                   **5** | Small categorical variable (e.g., New, Re-entry, etc.). Excellent for segmentation.                 |
| peak_date        |               **3,205** | Daily granularity.                                                                                  |
| entry_rank       |                 **195** | Entry rank values align with chart positions.                                                       |
| entry_date       |               **3,205** | Daily granularity.                                                                                  |
| release_date     |               **8,430** | Music released over approximately 23 years, enabling release-year trend analysis.                   |
| artist_uris      |              **97,427** | Unique Spotify artist identifiers. Better join key than artist names.                               |


1. Excellent Global Coverage

The dataset covers 71 countries, making it suitable for cross-country market analysis and regional trend identification.

2. Rich Music Catalog

With approximately:

256K unique songs
92K artists
28K labels

the dataset represents a highly diverse music ecosystem.

3. Long Historical Coverage

More than 3,200 daily chart snapshots allow:

Daily trend analysis
Monthly growth analysis
Seasonal pattern detection
Year-over-year comparisons
4. Strong Analytical Potential

High-cardinality identifiers (uri, artist_uris) support robust joins and entity-level analytics, while low-cardinality fields like entry_status are ideal for segmentation and dashboard filters.

In [37]:
#Step4- Descriptive Statistics(numerical Features)
numeric_cols = [
    "rank",
    "peak_rank",
    "previous_rank",
    "days_on_chart",
    "streams",
    "consecutive_days",
    "entry_rank"
]

charts_songs_df.select(numeric_cols).describe().show()

+-------+------------------+------------------+-----------------+------------------+-----------------+------------------+------------------+
|summary|              rank|         peak_rank|    previous_rank|     days_on_chart|          streams|  consecutive_days|        entry_rank|
+-------+------------------+------------------+-----------------+------------------+-----------------+------------------+------------------+
|  count|          42757018|          42757018|         42757018|          42757018|         42757018|          42757018|          42757018|
|   mean| 98.18277088453642| 34.44040110561499|87.21173822271703| 266.2623929245954|67799.69933520621|129.18163677831788|102.53703714791335|
| stddev|57.728287623528935|39.414182832340046|58.59409690234472|358.86676646748515|243949.6212990545|206.78136576272598| 70.07538717091514|
|    min|                 1|                 1|               -1|                 1|             1001|                 1|                 1|
|    max|    

In [38]:
# step5- Quartile Analysis
# Quartiles for Streams

quantiles = charts_songs_df.approxQuantile(
    "streams",
    [0.25, 0.50, 0.75],
    0.01
)

print("=" * 60)
print("Quartile Analysis - Streams")
print("=" * 60)
print("Q1 (25%)   :", quantiles[0])
print("Median     :", quantiles[1])
print("Q3 (75%)   :", quantiles[2])

Quartile Analysis - Streams
('Q1 (25%)   :', 4875.0)
('Median     :', 12453.0)
('Q3 (75%)   :', 44771.0)

1. Right-Skewed Distribution

Notice the large jump:

Q1      = 4,875
Median  = 12,453
Q3      = 44,771
Mean    = 67,800
Maximum = 30,987,370

The mean (67,800) is much higher than the median (12,453).

This indicates that a relatively small number of songs receive exceptionally high stream counts, pulling the average upward.

This is typical of music streaming platforms, where a few blockbuster tracks account for a large share of total streams.

2. Median vs Mean

Half of the chart entries receive fewer than approximately 12,453 streams.

Only a small fraction of songs generate hundreds of thousands or millions of streams.

3. Upper Quartile

The top 25% of chart entries receive more than 44,771 streams.

This threshold helps identify highly popular songs.

The Spotify streaming ecosystem follows a long-tail popularity distribution.

Most songs achieve moderate streaming volumes, while a relatively small number become viral or global hits, contributing disproportionately to overall platform streams.

This is expected behavior rather than a data quality issue.

In [39]:
#using IQR method for outlier detection
Q1 = 4875.0
Q3 = 44771.0

IQR = Q3 - Q1

lower_bound = Q1 - (1.5 * IQR)
upper_bound = Q3 + (1.5 * IQR)

print("=" * 60)
print("Streams Outlier Threshold")
print("=" * 60)
print("IQR          :", IQR)
print("Lower Bound  :", lower_bound)
print("Upper Bound  :", upper_bound)

Streams Outlier Threshold
('IQR          :', 39896.0)
('Lower Bound  :', -54969.0)
('Upper Bound  :', 104615.0)

In [40]:
from pyspark.sql import functions as F

upper_bound = 104615

outlier_count = charts_songs_df.filter(
    F.col("streams") > upper_bound
).count()

total_rows = charts_songs_df.count()

outlier_percent = (outlier_count * 100.0) / total_rows

print("=" * 60)
print("Streams Outlier Analysis")
print("=" * 60)
print("Outlier Records :", outlier_count)
print("Outlier %       :", outlier_percent)

Streams Outlier Analysis
('Outlier Records :', 5057340)
('Outlier %       :', 11.828093343647117)

These records should not be considered erroneous data.

Spotify streaming data naturally follows a long-tail distribution, where a relatively small proportion of songs generate exceptionally high streaming volumes due to:

Viral popularity
Playlist recommendations
New album releases
Seasonal songs
Global chart-topping hits
Social media trends

Therefore, these observations represent high-performing songs rather than data quality issues.

In [41]:
# correlation analysis
numeric_cols = [
    "rank",
    "peak_rank",
    "previous_rank",
    "days_on_chart",
    "streams",
    "consecutive_days",
    "entry_rank"
]

print("=" * 90)
print("Correlation Analysis")
print("=" * 90)

for i in range(len(numeric_cols)):
    for j in range(i + 1, len(numeric_cols)):

        col1 = numeric_cols[i]
        col2 = numeric_cols[j]

        corr_value = charts_songs_df.stat.corr(col1, col2)

        print(col1.ljust(20), "<->", col2.ljust(20), ":", corr_value,)

Correlation Analysis
('rank                ', '<->', 'peak_rank           ', ':', 0.4668591869420617)
('rank                ', '<->', 'previous_rank       ', ':', 0.7206922640833608)
('rank                ', '<->', 'days_on_chart       ', ':', 0.14351419850481742)
('rank                ', '<->', 'streams             ', ':', -0.128693204527507)
('rank                ', '<->', 'consecutive_days    ', ':', -0.14381361650068708)
('rank                ', '<->', 'entry_rank          ', ':', 0.21495961747801612)
('peak_rank           ', '<->', 'previous_rank       ', ':', 0.2889992868139932)
('peak_rank           ', '<->', 'days_on_chart       ', ':', -0.24179635096258345)
('peak_rank           ', '<->', 'streams             ', ':', -0.06493538154260128)
('peak_rank           ', '<->', 'consecutive_days    ', ':', -0.28729670536711505)
('peak_rank           ', '<->', 'entry_rank          ', ':', 0.5024182763486745)
('previous_rank       ', '<->', 'days_on_chart       ', ':', 0.159581344614400

| Variables                        | Correlation | Interpretation     |
| -------------------------------- | ----------: | ------------------ |
| Rank ↔ Previous Rank             |   **0.721** | Strong Positive    |
| Days on Chart ↔ Consecutive Days |   **0.504** | Moderate Positive  |
| Peak Rank ↔ Entry Rank           |   **0.502** | Moderate Positive  |
| Rank ↔ Peak Rank                 |   **0.467** | Moderate Positive  |
| Rank ↔ Streams                   |  **-0.129** | Very Weak Negative |
| Streams ↔ Days on Chart          |  **-0.013** | Negligible         |
| Streams ↔ Consecutive Days       |   **0.006** | Negligible         |


In [2]:
import sys

print("=" * 60)
print("Python Version")
print("=" * 60)
print(sys.version)

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Python Version
3.9.25 (main, May 25 2026, 00:00:00) 
[GCC 11.5.0 20240719 (Red Hat 11.5.0-5)]

In [3]:
print("=" * 60)
print("Spark Version")
print("=" * 60)
print(spark.version)

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Spark Version
3.5.0-amzn-0

In [4]:
print("=" * 60)
print("Hadoop Version")
print("=" * 60)

print(
    spark.sparkContext._jvm.org.apache.hadoop.util.VersionInfo.getVersion()
)

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Hadoop Version
3.3.6-amzn-2

In [5]:
print("=" * 60)
print("Default Parallelism")
print("=" * 60)

print(spark.sparkContext.defaultParallelism)

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Default Parallelism
2

In [6]:
for item in spark.sparkContext.getConf().getAll():
    print(item)

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

('spark.eventLog.enabled', 'true')
('spark.driver.extraLibraryPath', '/usr/lib/hadoop/lib/native:/usr/lib/hadoop-lzo/lib/native:/usr/lib/jvm/java-17-amazon-corretto.x86_64/lib/server:/docker/usr/lib/hadoop/lib/native:/docker/usr/lib/hadoop-lzo/lib/native:/docker/usr/lib/jvm/java-17-amazon-corretto.x86_64/lib/server')
('spark.driver.port', '38485')
('spark.sql.parquet.output.committer.class', 'com.amazon.emr.committer.EmrOptimizedSparkSqlParquetOutputCommitter')
('spark.blacklist.decommissioning.timeout', '1h')
('spark.yarn.appMasterEnv.SPARK_PUBLIC_DNS', '$(hostname -f)')
('spark.sql.emr.internal.extensions', 'com.amazonaws.emr.spark.EmrSparkSessionExtensions')
('spark.org.apache.hadoop.yarn.server.webproxy.amfilter.AmIpFilter.param.PROXY_URI_BASES', 'http://ip-172-31-70-86.ec2.internal:20888/proxy/application_1784304077754_0001')
('spark.driver.extraClassPath', '/usr/lib/hadoop-lzo/lib/*:/usr/lib/hadoop/hadoop-aws.jar:/usr/share/aws/aws-java-sdk/*:/usr/share/aws/aws-java-sdk-v2/*:/usr

In [1]:
from pyspark.sql import functions as F
from pyspark.sql import types as T
from pyspark.sql.window import Window

Starting Spark application


ID,YARN Application ID,Kind,State,Spark UI,Driver log,User,Current session?
1,application_1784304077754_0002,pyspark,idle,Link,Link,None,✔


FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

SparkSession available as 'spark'.


FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [ ]:
import sys

print("=" * 70)
print("Environment Verification")
print("=" * 70)

print(f"Driver Python version : {sys.version}")
print(f"Spark version         : {spark.version}")
print(f"Hadoop version        : "
      f"{spark.sparkContext._jvm.org.apache.hadoop.util.VersionInfo.getVersion()}")

executor_python = (
    spark.sparkContext
         .parallelize([1], 1)
         .map(lambda x: sys.version)
         .collect()[0]
)

print(f"Executor Python       : {executor_python}")
print(f"Default parallelism   : {spark.sparkContext.defaultParallelism}")
print(f"Dynamic allocation    : "
      f"{spark.conf.get('spark.dynamicAllocation.enabled', 'not set')}")

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [2]:
from pyspark.sql import functions as F
from pyspark.sql import types as T
from pyspark.sql.window import Window

spark.conf.set("spark.sql.adaptive.enabled", "true")
spark.conf.set("spark.sql.adaptive.coalescePartitions.enabled", "true")

print("Adaptive Query Execution:",
      spark.conf.get("spark.sql.adaptive.enabled"))

print("Adaptive Partition Coalescing:",
      spark.conf.get("spark.sql.adaptive.coalescePartitions.enabled"))

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Adaptive Query Execution: true
Adaptive Partition Coalescing: true